File: Programmers/[87694] 아이템 줍기/JFe/87694.py
from collections import deque

def bfs(field, characterX, characterY, itemX, itemY):
    visited = [[False for _ in range(101)]for _ in range(101)]
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]
    q = deque()
    q.append([characterX, characterY, 0])
    while q:
        x, y, v = q.popleft()
        visited[y][x] = True
        # itemX, itemY에 도착하면 종료
        if x == itemX and y == itemY: return v
        for i in range(4):
            nx, ny, nv = x + dx[i], y + dy[i], v + 1
            # 범위 안, 테두리 위, 방문 X면 BFS 탐색
            if 0 <= nx <= 100 and 0 <= ny <= 100 and field[ny][nx] and not visited[ny][nx]:
                q.append([nx, ny, nv])

def solution(rectangle, characterX, characterY, itemX, itemY):
    answer = 0
    field = [[False for _ in range(101)]for _ in range(101)]    # 크기 2배로 늘리기 (ㄷ자인 경우 처리하기 위함)
    # 직사각형 field에 표시
    for x1, y1, x2, y2 in rectangle:
        for i in range(y1*2, y2*2+1):
            for j in range(x1*2, x2*2+1):
                field[i][j] = True
    # 직사각형 내부 지우기
    for x1, y1, x2, y2 in rectangle:
        for i in range(y1*2+1, y2*2):
            for j in range(x1*2+1, x2*2):
                field[i][j] = False
    answer = bfs(field, characterX*2, characterY*2, itemX*2, itemY*2) / 2   # 2배로 늘렸기 때문에 2로 나눠서 return
    return answer

File: Programmers/[87694] 아이템 줍기/SUbbb/87694.java
import java.util.*;

class Coor {
    int x;
    int y;
    
    public Coor(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

class Solution {
    private static int[][] map = new int[101][101];
    private static int answer = Integer.MAX_VALUE;
    
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        for (int[] rect : rectangle)
            fillMap(rect);
        
        bfs(characterX * 2, characterY * 2, itemX * 2, itemY * 2);
        return answer/2;
    }
    
    // 주어진 사각형 정보를 이용해 테두리와 내부 정보를 map에 표시
    private void fillMap(int[] rect) {
        int x1 = rect[0] * 2;
        int y1 = rect[1] * 2;
        int x2 = rect[2] * 2;
        int y2 = rect[3] * 2;
        
        // 내부 색칠
        // -1 : 내부
        // 1 : 테두리
        for (int i = x1; i <= x2; i++) {
            for (int j = y1; j <= y2; j++) {
                if (map[i][j] == -1) continue;
                map[i][j] = -1;
                if (i == x1 || i == x2 || j == y1 || j == y2) map[i][j] = 1;
            }
        }       
    }
    
    private void bfs(int cX, int cY, int iX, int iY) {
        int[][] dir = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
        boolean[][] visited = new boolean[101][101];
        Queue<Coor> queue = new LinkedList<>();
        queue.add(new Coor(cX, cY));
        
        while(!queue.isEmpty()) {
            Coor c = queue.poll();
            int curX = c.x;
            int curY = c.y;
            
            for (int[] d : dir) {
                int nX = curX + d[0];
                int nY = curY + d[1];
                
                if (!isValid(nX, nY)) continue;
                if (map[nX][nY] != 1 || visited[nX][nY]) continue;
                
                // map에 비용 저장
                map[nX][nY] = map[curX][curY] + 1;
                if (nX == iX && nY == iY) {
                    // 이미 저장된 값이 있다면, 더 최소 값을 저장
                    answer = Math.min(answer, map[nX][nY]);
                    continue;
                }
                
                visited[nX][nY] = true;
                queue.add(new Coor(nX, nY));
            }
        }
    }
    
    private boolean isValid(int i, int j) {
        return i >= 0 && i < 101 && j >= 0 && j < 101;
    }
}